### DISCUSSION TITLE
poetry publish fails despite a dist folder being present

### DISCUSSION THREAD

**Discussion by tschm** [python-poetry/poetry#8357] [General] [answered]
Ok, I have been told it's better to keep the build and deployment in two separate jobs (for security reasons). So, I build my project (poetry build) and share the resulting dist folder with a second job that gets an ephemeral key and pushes to pypi. This used to work if the build step was in the same job. Here's an example:

https://github.com/tschm/mosek-license-server

**Comment by tschm** [comment]
i get

No files to publish. Run poetry build first or use the --build option.
Error: Process completed with exit code 1.

**Comment by tschm** [comment]
The dist folder is present and as expected packed with 

```
drwxr-xr-x  2 runner docker 4096 Aug 23 19:40 .
drwxr-xr-x 10 runner docker 4096 Aug 23 19:40 ..
-rw-r--r--  1 runner docker 7145 Aug 23 19:40 mosek_license_server-0.8.4-py3-none-any.whl
-rw-r--r--  1 runner docker 6168 Aug 23 19:40 mosek_license_server-0.8.4.tar.gz
No files to publish. Run poetry build first or use the --build option.
Error: Process completed with exit code 1.
```

**Reply by branchv** [reply, marked as answer]
The error message could definitely be improved, but `poetry` will only look for a built distribution matching the version number in `pyproject.toml`. Since your `build` job [updates this version](https://github.com/tschm/mosek-license-server/actions/runs/5955728287/job/16154975842#step:4:14), your `deploy` job will need to do the same

**Comment by geudrik** [comment]
Ayyup, I too fell for this tarp. That error message should really be updated to include this word of warning <3

**Comment by n-badtke-cg** [comment]
Ouch. I just fell over this error message.. and now I am not sure how to solve it. 
We have the version in the `__init__.py`:

```python
__version__ = "1.0.0"

## This is needed 
def get_info():
    return {
        "package-name": "some_package",  # Required
        "name": "some",  # Required
        "description": "A package for something.",  # Required
        "connection-types": [
            {
                "connection-type": "foo",
                "hook-class-name": "...."
            }
        ],
        "versions": [__version__],  # Required
    }
```

Then in the `pyproject.toml`, we are using setuptools for parsing the `__version__`:

```python
[build-system]
requires = [
    "setuptools",
    "wheel"
]
build-backend = "setuptools.build_meta"

[project]
dynamic = [
    "version"
]

# required by poetry but unused due to usage of setuptools
[tool.poetry]
version = "0"

[tool.setuptools.dynamic]
version = { attr = "some_package.__version__" }
```

`project.version` is "0", but `__version__` is the actual used value. 

Always needing to update both files is not [DRY](https://en.wikipedia.org/wiki/Don't_repeat_yourself).

Has anybody any ideas?
